### PR TITLE
Fix for process getting stuck while processing RC backups

### PR DESCRIPTION
**Pull request by jyotimisramsft**
Work item: https://dev.azure.com/msazure/One/_workitems/edit/24409835/
1. Updating to concurrent dictionary
2. Updating the code to use user provided log object

Ref: https://portal.microsofticm.com/imp/v3/incidents/details/387104335/home

### CODE REVIEW

File: src/BackupExplorer/Parser/BackupParserImpl.cs [hunk 1]
```diff
@@ -51,6 +51,39 @@ public BackupParserImpl(string backupChainPath, string codePackagePath)
...
+                new Uri("fabric:/rcbackupapp/rcbackupservice"),
+                DateTime.UtcNow.ToFileTimeUtc(),
+                DateTime.UtcNow.ToFileTimeUtc(),
+                this.OnDataLossCallback,
+                this.CreateStateProvider);
+            this.reliabilitySimulator.CreateReplica(true, false);
+
+            this.stateManager = new StateManager(reliabilitySimulator);
+            this.seenFirstTransaction = false;
+            this.transactionChangeManager = new TransactionChangeManager(this.reliabilitySimulator);
```
Comment (anantshankar17) [inline src/BackupExplorer/Parser/BackupParserImpl.cs:84]: Instead of duplicating all this can we just call the other method which does it already ?

**Review by anantshankar17 — CHANGES_REQUESTED** [review]

File: src/BackupExplorer/Parser/BackupParserImpl.cs (hunk 1 above)
Comment (jyotimisramsft) [reply src/BackupExplorer/Parser/BackupParserImpl.cs]: Updated to make it more modular

**Review by roshanroutray — CHANGES_REQUESTED** [review]
Added review comments. Please resolve.

File: src/BackupExplorer/Parser/BackupParser.cs [hunk 2]
```diff
@@ -35,6 +35,35 @@ public BackupParser(string backupChainPath)
...
 
+        /// <summary>
+        /// Constructor for BackupParser.
+        /// </summary>
+        /// <param name="backupChainPath">Folder path that contains sub folders of one full and multiple incremental backups.</param>
+        /// <param name="userLog">User configurable log object. </param>
+        /// <param name="workFolderPath">Folder pah for work folder. </param>
+        /// <param name="checkpointThreshold">Configurable checkpointThreshold. </param>
+        /// Pass an empty string if code package is not required for backup parsing. e.g. when backup has only primitive types.        
+        public BackupParser(string backupChainPath, ILog userLog, string workFolderPath, int checkpointThreshold = 200)
```
Comment (anantshankar17) [inline src/BackupExplorer/Parser/BackupParser.cs:46]: NIT: checkpointThresholdInMB.

File: src/BackupExplorer/Parser/BackupParserImpl.cs [hunk 3]
```diff
@@ -33,9 +33,36 @@ public BackupParserImpl(string backupChainPath, string codePackagePath)
...
             this.codePackage = new CodePackageInfo(codePackagePath);
             this.workFolder = Path.Combine(Directory.GetCurrentDirectory(), Guid.NewGuid().ToString());
+            this.checkpointThreshold = 200;
 
             Console.WriteLine("Work Folder : {0}", this.workFolder);
+            InitializeBackupParser();
+           
+        }
+
+        public BackupParserImpl(string backupChainPath, string codePackagePath, ILog log, string workFolderPath, int checkpointThreshold = 200)
```
Comment (anantshankar17) [inline src/BackupExplorer/Parser/BackupParserImpl.cs:43]: can there be just 1 constructor with default/null values for other params ?

File: src/BackupExplorer/Parser/BackupParser.cs [hunk 4]
```diff
@@ -35,6 +35,35 @@ public BackupParser(string backupChainPath)
...
+        /// <summary>
+        /// Constructor for BackupParser.
+        /// </summary>
+        /// <param name="backupChainPath">Folder path that contains sub folders of one full and multiple incremental backups.</param>
+        /// <param name="codePackagePath">code package path.</param>
+        /// <param name="userLog">User configurable log object. </param>
+        /// <param name="workFolderPath">Folder path for work folder. </param>
+        /// <param name="checkpointThreshold">Configurable checkpointThreshold. </param>
+        /// Pass an empty string if code package is not required for backup parsing. e.g. when backup has only primitive types.        
+        public BackupParser(string backupChainPath, string codePackagePath, ILog userLog, string workFolderPath, int checkpointThreshold = 200)
```
Comment (anantshankar17) [inline src/BackupExplorer/Parser/BackupParser.cs:61]: Instead of defining new constructors, can you use defaults and have just 1 constructor ?

**Review by anantshankar17 — CHANGES_REQUESTED** [review]

File: src/BackupExplorer/Parser/BackupParserImpl.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyotimisramsft) [reply src/BackupExplorer/Parser/BackupParserImpl.cs]: good point. Updated the variable name

File: src/BackupExplorer/Parser/BackupParserImpl.cs [hunk 5]
```diff
@@ -33,9 +33,36 @@ public BackupParserImpl(string backupChainPath, string codePackagePath)
             this.backupChainPath = backupChainPath;
             this.codePackage = new CodePackageInfo(codePackagePath);
             this.workFolder = Path.Combine(Directory.GetCurrentDirectory(), Guid.NewGuid().ToString());
+            this.checkpointThreshold = 200;
 
             Console.WriteLine("Work Folder : {0}", this.workFolder);
```
[the remark this answers is not in the document]
Comment (jyotimisramsft) [reply src/BackupExplorer/Parser/BackupParserImpl.cs]: Simplified to a single constructor with defaults

File: src/BackupExplorer/Parser/BackupParser.cs (hunk 2 above)
Comment (jyotimisramsft) [reply src/BackupExplorer/Parser/BackupParser.cs]: done.

File: src/BackupExplorer/Parser/BackupParserImpl.cs (hunk 3 above)
Comment (jyotimisramsft) [reply src/BackupExplorer/Parser/BackupParserImpl.cs]: Simplified to a single constructor with defaults

File: src/BackupExplorer/Parser/BackupParser.cs (hunk 4 above)
Comment (jyotimisramsft) [reply src/BackupExplorer/Parser/BackupParser.cs]: updated

**Review by roshanroutray — APPROVED** [review]

**Review by anantshankar17 — APPROVED** [review]

**Pull request merged**